### PR TITLE
Add flatten function for nested list

### PR DESCRIPTION
**Pull request by usurai**
Implements function `flatten` requested in #2922: `flatten(list<list<T>>) -> list<T>`.
The function takes a nested list, unnests it and concatenate the elements.
`flatten(DISTINCT list...)` is not yet implemented as I have some concern in terms of design. Will discuss in the issue thread.

### CODE REVIEW

File: test/sql/function/list/flatten.test [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# name: test/sql/function/list/flatten.test
+# description: Test flatten function
+# group: [list]
+
+statement ok
+PRAGMA enable_verification
+
+# basic functionality
+query T
+SELECT flatten([[1, 2], [3, 4]])
```
Comment (Mytherin) [inline test/sql/function/list/flatten.test:10]: Could we add a test with only empty lists (e.g. `select flatten([[], [], []])`?

File: test/sql/function/list/flatten.test [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+----
+[[1, 2], [3, 4], [5, 6], [7, 8]]
+
+query T
+SELECT flatten([[[1, 2], [3, 4]], [], [[5, 6], [7, 8]]]);
+----
+[[1, 2], [3, 4], [5, 6], [7, 8]]
+
+query T
+SELECT flatten([[[1, 2], [3, 4]], [[]], [[5, 6], [7, 8]]]);
```
Comment (Mytherin) [inline test/sql/function/list/flatten.test:45]: Could we add a test where we call flatten multiple times in a row?

File: test/sql/function/list/flatten.test [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+create table test as select range % 8 i, range j from range(16)
+
+statement ok
+create table lists as select i%4 i, list(j) j from test group by i
+
+statement ok
+create table list_lists as select i, list(j) j from lists group by i
+
+query II
+SELECT i, flatten(j) FROM list_lists
```
Comment (Mytherin) [inline test/sql/function/list/flatten.test:97]: Could we add a test where we add a filter here, e.g:

```sql
SELECT i, flatten(j) FROM list_lists WHERE i%2!=0
```

As well as a test where we flatten a constant vector several times, e.g.:

```sql
SELECT i, flatten([1, 2, 3]) FROM list_lists
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Could you have a look at resolving the merge conflict? Otherwise, some requests for more tests:

**Comment by usurai** [comment]
@Mytherin Thanks for the comments! Seems I messed up `NULL` and `[]` in the first take. Have fixed this and added tests as requested. Please take a look.

**Review by Mytherin — APPROVED** [review]
Thanks for the changes! Looks good.

**Comment by usurai** [comment]
Used release build to do the test and D_ASSERT did not trigger. Pushed new fix.

**Comment by Mytherin** [comment]
Could you have a look at the breaking vector_sizes test? You can test this locally by changing the `STANDARD_VECTOR_SIZE` to 2 in `duckdb/common/vector_size.hpp`.

**Comment by usurai** [comment]
> Could you have a look at the breaking vector_sizes test? You can test this locally by changing the `STANDARD_VECTOR_SIZE` to 2 in `duckdb/common/vector_size.hpp`.

Sorry for the late response. It took some time to figure out.
Turns out my method was incorrect when argument is dictionary vector: I made result to reference argument's child entry and assigned result's list entries to point to the flattened data. This works fine when the argument is flat / constant vector. But when the argument is dictionary vector which has some data not selected, the result's aux vector, which is a flat vector, will contain unneeded data.
The fix is to change my method to copy needed data to the result, instead of referencing the whole vector. This change passes vector size test in local.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**